### PR TITLE
Adds handling for ListenToInternalEvents on eventProvider

### PR DESCRIPTION
**Pull request by haeferer**
This includes

* added Attribute in EventProvider and Filteroptions
* rework publishEvents to support publishOptions
* some changes from prettier-run !?

### CODE REVIEW

**Comment by jlkalberer** [comment]
@AntonPuko

File: src/lib/EventPublisher.js [hunk 1]
```diff
@@ -77,7 +81,9 @@ class EventPublisher extends EventEmitter {
     eventData: EventData,
   ): Promise<Object> => {
     const eventID = uuid();
-    const requestEventName = `${getRequestEventName(eventData.name)}/${eventID}`;
+    const requestEventName = `${getRequestEventName(
```
Comment (AntonPuko) [inline src/lib/EventPublisher.js:85]: this formatting can't be prettier, I just tried to commit this line with last dev and it fixes to one line.

File: src/lib/EventPublisher.js [hunk 2]
```diff
@@ -140,25 +151,30 @@ class EventPublisher extends EventEmitter {
...
-      this.once(eventNamePrefix, (): void => clearTimeout(timeout));
+      this.once(eventNamePrefix, () => {
+        clearTimeout(timeout);
+      });
     }
 
     if (once) {
       this.once(eventNamePrefix, listener);
+      this.once(eventNamePrefix, () => {
+        this._subscriptionsByID.delete(subscriptionID);
```
Comment (AntonPuko) [inline src/lib/EventPublisher.js]: you should call listener with the event here. man...why the same mistakes....? Did you test it? IF yes, it shouldn't work. and actually i don't understand how it passes the existing tests. hm. Actually, I just tested, the test doesn't pass withoout calll the listener()...so how did you even commit this?

File: src/lib/EventPublisher.js [hunk 3]
```diff
@@ -140,25 +151,30 @@ class EventPublisher extends EventEmitter {
           timeoutHandler();
         }
       }, subscriptionTimeout);
-
-      this.once(eventNamePrefix, (): void => clearTimeout(timeout));
+      this.once(eventNamePrefix, () => {
```
Comment (AntonPuko) [inline src/lib/EventPublisher.js]: why did you add parenthesis here?

File: src/lib/EventPublisher.js [hunk 4]
```diff
@@ -140,25 +151,30 @@ class EventPublisher extends EventEmitter {
...
   unsubscribe = (subscriptionID: string) => {
-    const { eventNamePrefix, listener } = nullthrows(
-      this._subscriptionsByID.get(subscriptionID),
+    const subscription: Subscription = this._subscriptionsByID.get(
+      subscriptionID,
     );
-
-    this.removeListener(eventNamePrefix, listener);
+    if (!subscription) {
+      return;
```
Comment (AntonPuko) [inline src/lib/EventPublisher.js:174]: are you sure this return didn't add a Flow error?

**Comment by AntonPuko** [comment]
write one more simple test in test/EventPublisher.js for the filtering you adding.

File: src/lib/EventPublisher.js (hunk 3 above)
Comment (haeferer) [reply src/lib/EventPublisher.js]: It looks more natural too me. Thats all, but i will change this

File: src/lib/EventPublisher.js (hunk 4 above)
Comment (haeferer) [reply src/lib/EventPublisher.js:174]: i cant interpret the error, i just took you suggestion to deal with already remove subscriptions.
maybe you can help me to understand what's wrong
```
const subscription: Subscription = this._subscriptionsByID.get(
170:       subscriptionID,
171:     );
         ^ undefined. This type is incompatible with
169:     const subscription: Subscription = this._subscriptionsByID.get(
                             ^^^^^^^^^^^^ object type
```

File: src/lib/EventPublisher.js (hunk 3 above)
Comment (AntonPuko) [reply src/lib/EventPublisher.js]: Try don't change project codestyle just because you like something more.

File: src/lib/EventPublisher.js (hunk 2 above)
Comment (haeferer) [reply src/lib/EventPublisher.js]: Help me to understand the problem:

in case of once, there are 2 functions subscribed which will be remove automatically (once). first is the listener, second is removing the subscribtion...

What's wrong with this

```
this.once(eventNamePrefix, listener);
this.once(eventNamePrefix, () => {
        this._subscriptionsByID.delete(subscriptionID);
});
```

File: src/lib/EventPublisher.js (hunk 4 above)
Comment (AntonPuko) [reply src/lib/EventPublisher.js:174]: 1) you shouldn't explicitly cast `subscription: Subscription` Flow is smart enough to understand the type from the map.
2) Your casting is wrong because it's `?Subscription`

File: src/lib/EventPublisher.js (hunk 2 above)
Comment (AntonPuko) [reply src/lib/EventPublisher.js]: ah..you has one more subscription on line before..ok sorry..but rather do it in one function.

File: src/lib/EventPublisher.js (hunk 4 above)
Comment (haeferer) [reply src/lib/EventPublisher.js:174]: got it

File: src/lib/EventPublisher.js (hunk 2 above)
Comment (haeferer) [reply src/lib/EventPublisher.js]: I cant lint right now cause 
````7:22  error  Unable to resolve path to module '../../third-party/settings.json'   import/no-unresolved
  8:28  error  Unable to resolve path to module '../../third-party/specifications'  import/no-unresolved
  8:28  error  Missing file extension for "../../third-party/specifications"        import/extensions
  9:22  error  Unable to resolve path to module '../../third-party/versions.json'   import/no-unresolved
```

Comment (haeferer) [reply src/lib/EventPublisher.js]: Ok, cause of these lint errors i cant commit at all. 

so that's all for today :(

Comment (haeferer) [reply src/lib/EventPublisher.js]: It's working now (problem is npm install doesnt run well on MacOS).
I've changed the code

File: src/lib/EventPublisher.js (hunk 3 above)
Comment (haeferer) [reply src/lib/EventPublisher.js]: done

**Comment by haeferer** [comment]
added a test, TEST LINT BUILD and FLOW lock ok

finished from my side

**Comment by haeferer** [comment]
After this PR i will make a PR for BunyanLogger for spark-protocol. Work is finished.

File: src/lib/EventPublisher.js [hunk 5]
```diff
@@ -56,13 +57,17 @@ type Subscription = {
 class EventPublisher extends EventEmitter {
   _subscriptionsByID: Map<string, Subscription> = new Map();
 
-  publish = (eventData: EventData) => {
-    const ttl = eventData.ttl && eventData.ttl > 0
-      ? eventData.ttl
-      : settings.DEFAULT_EVENT_TTL;
+  publish = (eventData: EventData, options: PublishOptions) => {
+    const { isInternal, isPublic } = options || {};
```
Comment (AntonPuko) [inline src/lib/EventPublisher.js:61]: this incorrect: 
1) if `options: PublishOptions`, we should always have them in args and `|| {}` becomes useless. if not, you should change the type.
2) do destruction inside function args.

File: test/EventPublisher.test.js [hunk 6]
```diff
@@ -9,302 +9,264 @@ import { getRequestEventName } from '../src/lib/EventPublisher';
...
+  };
+
+  eventPublisher.subscribe(eventData.name, handler, {
+    filterOptions: { listenToInternalEvents: true },
+  });
+
+  eventPublisher.publish(eventData);
+
+  process.nextTick(() => {
+    t.falsy(handler.called);
```
Comment (AntonPuko) [inline test/EventPublisher.test.js:86]: this doesn't make sense. if you do `filterOptions: { listenToInternalEvents: true }`  handler should be called, so:
1) filter means, the event should be filtered, so you need to change to: `{ listenToInternalEvents: false }`
2) if current test passes: there is a bug in the code, if not: your env didn't run the tests before commit and you was able to commit without tests check.

Comment (haeferer) [reply test/EventPublisher.test.js:86]: I'm currently rewriting your Testcode, because i think all tests including eventhandling don't work (the way they are designed), they are simply true (always). I will comment later on this.

Comment (jlkalberer) [reply test/EventPublisher.test.js:86]: Hey.. You should do multiple PRs instead of making a giant PR.

If the tests should be rewritten, it should be done in separate PR so it's easier to review.

Comment (AntonPuko) [reply test/EventPublisher.test.js:86]: ok, if all the tests are wrong, lets merge this. Thanks for the work.

Comment (haeferer) [reply test/EventPublisher.test.js:86]: i will start a new PR  with this, but first will be bunyan-log for spark-protocol so we can continue with bunyan-log on spark-server

Comment (AntonPuko) [inline test/EventPublisher.test.js]: seems, tests are going bad just because of nextTick on them.

Comment (haeferer) [reply test/EventPublisher.test.js:86]: No nextTick is needed to get the event done, nut you cant wait on NEXTTICK...!  
I think you have to return a promise (so ava waits for the result).
Right now you are returning without exeception, and mark the test "as false" later (too late). 
I dont think ava can handle this strategy. If the test doesnt return a promise, the test is marked as "passed"

Comment (haeferer) [reply test/EventPublisher.test.js:86]: I Will design a TestStub for this

Comment (AntonPuko) [reply test/EventPublisher.test.js:86]: ah, yeah, makes sense
